src/legacy/core_plugins/tile_map/public/components/wms_options.tsx

```diff
@@ -0,0 +1,106 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React, { useEffect, useState } from 'react';
+import { EuiPanel, EuiSpacer, EuiTitle } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+
+import { toastNotifications } from 'ui/notify';
+import { VisOptionsSetValue } from 'ui/vis/editors/default';
+import { TmsLayer } from 'ui/vis/map/service_settings';
+import { ExtendedVisOptionsProps } from '../../../kbn_vislib_vis_types/public/utils/with_injected_dependencies';
+import { SelectOption } from '../../../kbn_vislib_vis_types/public/controls/select';
+import { SwitchOption } from '../../../kbn_vislib_vis_types/public/controls/switch';
+import { WmsInternalOptions } from './wms_internal_options';
+
+const mapLayerForOption = ({ id }: TmsLayer) => ({ text: id });
+
+function WmsOptions({ serviceSettings, stateParams, setValue, vis }: ExtendedVisOptionsProps) {
+  const { wms } = stateParams;
+  const { tmsLayers } = vis.type.editorConfig.collections;
+  const [tmsLayerOptions, setTmsLayersOptions] = useState(
+    vis.type.editorConfig.collections.tmsLayers.map(mapLayerForOption)
+  );
+
+  const setWmsOption: VisOptionsSetValue = (paramName, value) =>
+    setValue('wms', {
+      ...wms,
+      [paramName]: value,
+    });
+
+  useEffect(() => {
+    serviceSettings
+      .getTMSServices()
+      .then(services => {
+        const newBaseLayers = [
+          ...tmsLayers,
+          ...services.filter(service => !tmsLayers.some(({ id }: TmsLayer) => service.id === id)),
+        ];
+
+        setTmsLayersOptions(newBaseLayers.map(mapLayerForOption));
+
+        if (!wms.selectedTmsLayer && newBaseLayers.length) {
+          setWmsOption('selectedTmsLayer', newBaseLayers[0]);
+        }
+      })
+      .catch((error: Error) => toastNotifications.addWarning(error.message));
+  }, []);
+
+  return (
+    <EuiPanel paddingSize="s">
+      <EuiTitle size="xs">
+        <div>
+          <FormattedMessage
+            id="tileMap.wmsOptions.baseLayerSettingsTitle"
+            defaultMessage="Base layer settings"
+          />
+        </div>
+      </EuiTitle>
+      <EuiSpacer size="s" />
+      {!wms.enabled && (
+        <SelectOption
+          label={i18n.translate('tileMap.wmsOptions.layersLabel', {
+            defaultMessage: 'Layers',
+          })}
+          options={tmsLayerOptions}
+          paramName="selectedTmsLayer"
+          value={wms.selectedTmsLayer && wms.selectedTmsLayer.id}
+          setValue={setWmsOption}
+        />
+      )}
+
+      <SwitchOption
```

Comment:
The only other thing I'd change is that since this Switch changes whether the Layer selection box is available, I'd move the switch to be **above** the layer select.

<img src="https://d.pr/free/i/LiHh1H+" width="50%" />

Reply:
done